Add "Key-Value Data" and "SQL" architecture cookbook recipes (#11410)
flutter/website: src/content/app-architecture/cookbook/key-value-data.md

---
title: "Persistent storage architecture: Key-value data"
description: Create a service to store key-value data
js:
  - defer: true
    url: /assets/js/inject_dartpad.js
---

<?code-excerpt path-base="app-architecture/todo_data_service"?>

Most Flutter applications, no matter how small or big they are,
require storing data on the user’s device at some point, such as API keys, 
user preferences or data that should be available offline.

In this recipe, you will learn how to integrate persistent storage 
for key-value data in a Flutter application 
that uses the recommended [Flutter architecture design][]. 
If you aren’t familiar with storing data to disk at all, 
you can read the [Store key-value data on disk][] recipe. 

Key-value stores are often used for saving simple data, 
such as app configuration, 
and in this recipe you’ll use it to save Dark Mode preferences. 

> If you want to learn how to store complex data on device, 

Reviewer: ```suggestion
If you want to learn how to store complex data on a device, 
```

    toggle = Command0(_toggle);
  }

  final ThemeRepository _themeRepository;

  bool _isDarkMode = false;

  /// If true show dark mode
  bool get isDarkMode => _isDarkMode;

  late Command0 load;

  late Command0 toggle;

  /// Load the current theme setting from the repository
  Future<Result<void>> _load() async {
    try {
      final result = await _themeRepository.isDarkMode();
      if (result is Ok<bool>) {
        _isDarkMode = result.value;
      }
      return result;
    } on Exception catch (e) {
      return Result.error(e);
    } finally {
      notifyListeners();
    }
  }

  /// Toggle the theme setting
  Future<Result<void>> _toggle() async {
    try {
      _isDarkMode = !_isDarkMode;
      return await _themeRepository.setDarkMode(_isDarkMode);
    } on Exception catch (e) {
      return Result.error(e);
    } finally {
      notifyListeners();
    }
  }
}
```

### Theme selection data layer

Following the architecture guidelines, 
the data layer is split into two parts: 
the `ThemeRepository` and the `SharedPreferencesService`.

The `ThemeRepository` is the single source of truth 
for all the theming configuration settings, 

> and will handle any possible errors coming from the service layer.

Reviewer: ```suggestion
and handles any possible errors coming from the service layer.
```

In this example, 
the `ThemeRepository` also exposes the dark mode setting 

> through an observable Stream. 

Reviewer: ```suggestion
through an observable `Stream`. 
```

This allows other parts of the application 
to subscribe to changes in the dark mode setting.

> The `ThemeRepository` depends on the `SharedPreferencesService`.

Reviewer: ```suggestion
The `ThemeRepository` depends on `SharedPreferencesService`.
```

> The repository will obtain the stored value from the service, 

Reviewer: ```suggestion
The repository obtains the stored value from the service, 
```

> and will store it when it changes.

Reviewer: ```suggestion
and stores it when it changes.
```

The `setDarkMode()` method passes the new value to the `StreamController`,
so that any component listening to the `observeDarkMode` stream 

> will receive updates.

Reviewer: ```suggestion receives updates.
```

to store the dark mode setting, 
hiding this third-party dependency from the rest of the application

:::note
A third-party dependency is a way to refer to packages and plugins 
eveloped by other developers outside your organization.
:::

<?code-excerpt "lib/data/services/shared_preferences_service.dart (SharedPreferencesService)"?>
```dart
class SharedPreferencesService {
  static const String _kDartMode = 'darkMode';

  Future<void> setDarkMode(bool value) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_kDartMode, value);
  }

  Future<bool> isDarkMode() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(_kDartMode) ?? false;
  }
}
```

## Putting it all together

In this example, 
the `ThemeRepository` and `SharedPreferencesService` are created 
in the `main()` method 
and passed to the `MainApp` as constructor argument dependency.

<?code-excerpt "lib/main.dart (MainTheme)"?>
```dart
void main() {
// ···
  runApp(
    MainApp(
      themeRepository: ThemeRepository(
        SharedPreferencesService(),
      ),
// ···
    ),
  );
}
```

Then, when the `ThemeSwitch` is created, 

> create as well the `ThemeSwitchViewModel` 

Reviewer: ```suggestion
also create `ThemeSwitchViewModel` 
```
